src/HooglePlus/FilterTest.hs

```diff
@@ -113,80 +110,81 @@ mapRandomMultiple env count = do
+-- generate test input as pair (type, value)
+generateTestInput' :: [String] -> FunctionSigniture -> IO [(String, String)]
+generateTestInput' imports (constraints, args, retType) = mapM f (transformType constraints args)
   where
     f x = let str = show x in do
-      val <- generateRandomValue str
+      val <- generateRandomValue imports str
       return (str, val)
 
     transformType [] args = args
     transformType (envItem:env) args = transformType env args'
       where
         args' = map (replace envItem) args
 
-        replace (Polymorphic name, cons) (Polymorphic name') =
-          if name == name' then
-            case cons of
+        -- replace all occurrance of nameL to its instantiated type
+        replace (Polymorphic nameL, constraint) (Polymorphic nameR) =
+          if nameL == nameR then
+            case constraint of
               "Eq" -> Concrete "Int"
               "Num" -> Concrete "Int"
-          else Polymorphic name'
-        replace cons (ArgTypeList argType) = ArgTypeList (replace cons argType)
+          else Polymorphic nameR
+        replace constraint (ArgTypeList argType) = ArgTypeList (replace constraint argType)
         replace _ x = x
 
 eval_ :: [String] -> String -> Int -> IO (Maybe (Either InterpreterError String))
-eval_ modules line time = timeout time $ runInterpreter $ do {
-  setImports modules;
+eval_ modules line time = timeout time $ runInterpreter $ do
+  setImports modules
   eval line
-}
 
 runChecks :: Environment -> RType -> UProgram -> IO Bool
 runChecks env goalType prog = do
   list <- replicateM 10 (runChecks' modules funcSig body)
   return $ or list
   where
-    args = _arguments env
-    modules = "Prelude" : Set.toList (_included_modules env)
-    argList = Map.toList args
-    argNames = map fst argList
-    argTypes = map snd argList
-    monoGoals = map toMonotype argTypes
-    funcSig = mkFunctionSigStr (monoGoals ++ [goalType])
-    body = mkLambdaStr argNames prog
+    (modules, funcSig, body, _) = extractSolution env goalType prog
 
 runChecks' :: [String] -> String -> String -> IO Bool
-runChecks' modules funcSig body = if filterInf body then handleNotSupported $ do
-  arg <- (mapRandomArguments . parseTypeString) funcSig
-  result <- eval_ modules (fmtFunction_ body funcSig arg) 3000000
-  case result of
-    Nothing -> putStrLn "Timeout in running always-fail detection" >> return True
-    Just (Left err) -> putStrLn (displayException err) >> return False
-    Just (Right res) -> putStrLn res >> return True
-else pure $ not rejectInfiniteFunctions
-
-filterInf :: String -> Bool
-filterInf body = not $ any (`isInfixOf` body) excludeFunctions
-
+runChecks' modules funcSig body =
+  if hasNoInfiniteFuncs body
+  then handleNotSupported executeCheck
+  else pure $ not rejectInfiniteFunctions
+  where
+    executeCheck = do
+      arg <- generateTestInput modules (parseTypeString funcSig)
+      result <- eval_ modules (fmtFunction_ body funcSig arg) 3000000
```

Comment:
The timeout number is still all around, could you move it to a constant somewhere?